src/main/java/com/commercetools/sync/customobjects/CustomObjectSync.java

```diff
@@ -0,0 +1,329 @@
+     * {@code validCustomObjectDrafts}, the matching custom objects in the target CTP project are fetched then the
+     * method {@link CustomObjectSync#syncBatch(Set, Set)} is called to perform the sync (<b>update</b> or <b>create</b>
+     * requests accordingly) on the target project.
+     *
+     * <p> In case of error during of fetching of existing custom objects, the error callback will be triggered.
+     * And the sync process would stop for the given batch.
+     * </p>
+     *
+     * @param batch batch of drafts that need to be synced
+     * @return a {@link CompletionStage} containing an instance
+     * of {@link CustomObjectSyncStatistics} which contains information about the result of syncing the supplied
+     * batch to the target project.
+     */
+
+    protected CompletionStage<CustomObjectSyncStatistics> processBatch(
+        @Nonnull final List<CustomObjectDraft<JsonNode>> batch) {
+
+        final Set<CustomObjectDraft<JsonNode>> validCustomObjectDrafts = batch.stream().filter(
+            this::validateDraft).collect(toSet());
+
+        if (validCustomObjectDrafts.isEmpty()) {
+            statistics.incrementProcessed(batch.size());
+            return completedFuture(statistics);
+        } else {
+            final Set<CustomObjectCompositeIdentifier> identifiers = validCustomObjectDrafts.stream().map(draft ->
+                CustomObjectCompositeIdentifier.of(draft)).collect(toSet());
+
+            return customObjectService
+                .fetchMatchingCustomObjects(identifiers)
+                .handle(ImmutablePair::new)
+                .thenCompose(fetchResponse -> {
+                    final Set<CustomObject<JsonNode>> fetchedCustomObjects = fetchResponse.getKey();
+                    final Throwable exception = fetchResponse.getValue();
+
+                    if (exception != null) {
+                        final String errorMessage = format(CTP_CUSTOM_OBJECT_FETCH_FAILED, identifiers);
+                        handleError(errorMessage, exception, identifiers.size());
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return syncBatch(fetchedCustomObjects, validCustomObjectDrafts);
+                    }
+                })
+                .thenApply(ignored -> {
+                    statistics.incrementProcessed(batch.size());
+                    return statistics;
+                });
+        }
+    }
+
+    /**
+     * Checks if a draft is valid for further processing. If so, then returns {@code true}. Otherwise handles an error
+     * and returns {@code false}. A valid draft is a {@link CustomObjectDraft} object that is not {@code null} and its
+     * key is not empty.
+     *
+     * @param draft nullable draft
+     * @return boolean that indicate if given {@code draft} is valid for sync
+     */
+    private boolean validateDraft(@Nullable final CustomObjectDraft draft) {
+        if (draft == null) {
```

Comment:
I checked `CustomObject` in SDK, key and container cannot be empty when draft is instantiated. Otherwise `IllegalArgumentException` is thrown during instantiation. That's why i make `validateDraft` simpler to check whether the draft is empty.

The javadoc has been changed.
https://github.com/commercetools/commercetools-sync-java/pull/569/commits/3df023286da098dd224a1531cef4d972493b4c8d

Reply:
I checked again, you are right for the null case, and I check now that constructor of `CustomObjectDraft` has the validated calls so we don't need to check again, that's good!

Reply:
I tried with: 

````java
final CustomObjectDraft<JsonNode> customObjectDraft =
            CustomObjectDraft.ofUnversionedUpsert("#a", "key", null);
````
It failed with;

````
The container "#a" does not have the correct format. Key and container need to match [-_~.a-zA-Z0-9]+.
java.lang.IllegalArgumentException: The container "#a" does not have the correct format. Key and container need to match [-_~.a-zA-Z0-9]+.
````